oligo_designer_toolsuite/database/_oligo_database.py

```diff
@@ -721,58 +575,140 @@ def get_sequence_oligoid_mapping(
+
+    def filter_database_by_region(self, remove_region: bool, region_ids: Union[str, List[str]]) -> None:
+
+        # Check formatting
+        region_ids = check_if_list(region_ids)
+        if self.database:
+            for region_id in self.database.keys():
+                if remove_region and (region_id in region_ids):
+                    del self.database[region_id]
+                elif not remove_region and (region_id not in region_ids):
+                    del self.database[region_id]
+        else:
+            raise ValueError(
+                "Can not filter. Database is empty! Call the method load_database() or load_database_from_fasta() first."
+            )
+
+    def filter_database_by_oligo(self, remove_region: bool, oligo_ids: Union[str, List[str]]) -> None:
+
+        # Check formatting
+        oligo_ids = check_if_list(oligo_ids)
+        if self.database:
+            for region_id in self.database.keys():
+                oligo_ids_region = list(self.database[region_id].keys())
+                for oligo_id in oligo_ids_region:
+                    if remove_region and (oligo_id in oligo_ids):
+                        del self.database[region_id][oligo_id]
+                    elif not remove_region and (oligo_id not in oligo_ids):
+                        del self.database[region_id][oligo_id]
+        else:
+            raise ValueError(
+                "Can not filter. Database is empty! Call the method load_database() or load_database_from_fasta() first."
+            )
+
+    def filter_database_by_attribute_threshold(
+        self, attribute_name: str, attribute_thr: float, remove_if_smaller_threshold: bool
+    ) -> None:
+        oligos_to_delete = []
+        for region_id in self.database.keys():
+            for oligo_id in self.database[region_id].keys():
+                attribute_values = check_if_list(
+                    self.get_oligo_attribute_value(
+                        attribute=attribute_name, region_id=region_id, oligo_id=oligo_id, flatten=True
+                    )
+                )
+                if attribute_values:
+                    if remove_if_smaller_threshold and any(item < attribute_thr for item in attribute_values):
+                        oligos_to_delete.append((region_id, oligo_id))
+                    elif not remove_if_smaller_threshold and all(
+                        item > attribute_thr for item in attribute_values
+                    ):
+                        oligos_to_delete.append((region_id, oligo_id))
+
+        for region_id, oligo_id in oligos_to_delete:
+            del self.database[region_id][oligo_id]
+
+    def filter_database_by_attribute_category(
+        self, attribute_name: str, attribute_category: Union[str, List[str]], remove_if_equals_category: bool
+    ) -> None:
+
+        # Check formatting
+        attribute_category = check_if_list(attribute_category)
+        oligos_to_delete = []
+
+        for region_id in self.database.keys():
```

Comment:
The code could be parallelized region-wise.

Reply:
True, we will address it in the oligo attribute issue